### PR TITLE
BAQE-1239 Develop task assignment batch load test

### PR DESCRIPTION
**Pull request by rsynek**
During a development I added also the "Update" scenario, which is tracked under BAQE-1349. Sorry for mixing it in the same PR.

### CODE REVIEW

File: jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            if (assignedTasks.isEmpty()) { // If there are no tasks to work on, wait and retry.
+                sleep(500L);
+                continue;
+            }
+            LOGGER.debug(String.format("Completing %d tasks. Already finished %d tasks of %d.",
+                    assignedTasks.size(), completedTasks.size(), taskCount));
+            // Run each task in a separate thread.
+            for (TaskInstance task : assignedTasks) {
+                // Avoid submitting task some other thread already takes care of
+                if (!tasksInProgress.contains(task.getId())) {
```
Comment (MarianMacik) [inline jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java:131]: Maybe this is just a precaution but given that you iterate over a particular task just once, can it even happen?

File: jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    /**
+     * Computes delays between completing and starting consecutive tasks.
+     */
+    private List<Long> delaysBetweenCompleteAndStartEvents(SortedSet<TaskEventInstance> events) {
+        List<Long> delays = new ArrayList<>();
+        TaskEventInstance previous = null;
+        for (TaskEventInstance taskEventInstance : events) {
+            if (TASK_COMPLETED.equals(taskEventInstance.getType())) {
+                previous = taskEventInstance;
+            } else if (previous != null && TASK_STARTED.equals(previous.getType())) {
```
Comment (MarianMacik) [inline jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java:170]: I guess here it should be `TASK_STARTED.equals(taskEventInstance.getType())`? As you never assign `TASK_STARTED` type to the previous.

File: jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/TaskAssigning.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    protected void abortAllProcesses() {
+        List<ProcessInstance> processInstanceList = processClient.findProcessInstances(CONTAINER_ID, 0, Integer.MAX_VALUE);
+        List<Long> processInstanceIdList = processInstanceList.stream()
+                .map(processInstance -> processInstance.getId())
+                .collect(Collectors.toList());
+        processClient.abortProcessInstances(CONTAINER_ID, processInstanceIdList);
+    }
+
+    protected Collection<CompletableFuture<Void>> startProcessesAsync(int processCount, int numberOfThreads) {
```
Comment (MarianMacik) [inline jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/TaskAssigning.java:84]: Do you want to keep that method now you don't use async start of processes?

File: jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/batch/TaskAssigningBatch.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    @Override
+    public void execute() {
+        beforeScenario();
+
+        scheduledExecutorService.scheduleAtFixedRate(() -> {
+            TaskQueryFilterSpec filterSpec = new TaskQueryFilterSpec();
+            List<TaskInstance> unassignedTasks = getQueryClient().findHumanTasksWithFilters(UNASSIGNED_TASKS_QUERY_NAME,
+                    filterSpec, 0, 100);
```
[the remark this answers is not in the document]
Comment (rsynek) [reply jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/batch/TaskAssigningBatch.java:91]: Here even 1 should be sufficient.

File: jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java (hunk 1 above)
Comment (rsynek) [reply jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java:131]: It's not just a precaution. Due to starting & completing tasks asynchronously, some tasks are not started before the next query in the outer loop on the line 120. As a result, such a task is retrieved again and once more submitted for start & complete, which fails as one of the threads working on the same task changes its state first.

File: jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/TaskAssigning.java (hunk 3 above)
Comment (rsynek) [reply jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/TaskAssigning.java:84]: No, if it's not used, no reason for it to stay.

File: jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java (hunk 2 above)
Comment (rsynek) [reply jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java:170]: Nice catch! Thank you.

**Comment by rsynek** [comment]
@MarianMacik thanks for a review, your comments have been addressed.

**Pull request merged**